src/Be.Vlaanderen.Basisregisters.MessageHandling.Kafka.Producer/Producer.cs

```diff
@@ -0,0 +1,133 @@
+namespace Be.Vlaanderen.Basisregisters.MessageHandling.Kafka.Producer
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Confluent.Kafka;
+    using Newtonsoft.Json;
+    using Offset = Kafka.Offset;
+
+    public sealed class Producer : IProducer, IDisposable
+    {
+        private readonly ProducerOptions _producerOptions;
+        private readonly IProducer<string, string> _producer;
+
+        internal Producer(ProducerOptions producerOptions, IProducer<string, string> producer)
+        {
+            _producerOptions = producerOptions;
+            _producer = producer;
+        }
+
+        public Producer(ProducerOptions producerOptions)
+         : this(
+             producerOptions,
+             new ProducerBuilder<string, string>(producerOptions.CreateProduceConfig())
+                .SetKeySerializer(Serializers.Utf8)
+                .SetValueSerializer(Serializers.Utf8)
+                .Build())
+        { }
+
+        public async Task<Result> Produce(
+            MessageKey key,
+            string message,
+            IEnumerable<MessageHeader>? headers = null,
+            CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                var offset = await ProduceMessage(key, message, headers, cancellationToken);
+
+                return Result.Success(offset);
+            }
+            catch (ProduceException<Null, string> ex)
+            {
+                return Result.Failure(ex.Error.Code.ToString(), ex.Error.Reason);
+            }
+            catch (OperationCanceledException ex)
+            {
+                return Result.Failure(ex.Message, ex.Message);
+            }
+        }
+
+        public async Task<Result> ProduceJsonMessage<T>(
+            MessageKey key,
+            T message,
+            IEnumerable<MessageHeader>? headers = null,
+            CancellationToken cancellationToken = default)
+            where T : class
+        {
+            try
+            {
+                var serializer = JsonSerializer.CreateDefault(_producerOptions.JsonSerializerSettings);
+                var kafkaJsonMessage = JsonMessage.Create(message, serializer);
+
+                var offset = await ProduceMessage(key, serializer.Serialize(kafkaJsonMessage), headers, cancellationToken);
+
+                return Result.Success(offset);
+            }
+            catch (ProduceException<Null, JsonMessage> ex)
+            {
+                return Result.Failure(ex.Error.Code.ToString(), ex.Error.Reason);
+            }
+            catch (OperationCanceledException)
+            {
+                return Result.Success(default);
```

Comment:
Just a question, not a comment. OperationCanceledException, is it thrown when idempotency is enabled?

Reply:
Just noticed that in the above Produce() we do: return Result.Failure(ex.Message, ex.Message); in case of this exception.

Reply:
good catch, gonna make them consistent